### PR TITLE
OverrideTip refactoring

### PR DESCRIPTION
**Pull request by blackyblack**

### CODE REVIEW

File: include/veriblock/pop/blockchain/pop/fork_resolution.hpp [hunk 1]
```diff
@@ -552,6 +552,7 @@ struct PopAwareForkResolutionComparator {
       }
 
       VBK_LOG_DEBUG("Candidate contains VALID commands, chain B wins");
+      ed_.overrideTip(candidate);
```
Comment (Warchant) [inline include/veriblock/pop/blockchain/pop/fork_resolution.hpp:555]: Why did you make this change?

File: include/veriblock/pop/blockchain/pop/fork_resolution.hpp [hunk 2]
```diff
@@ -662,6 +663,7 @@ struct PopAwareForkResolutionComparator {
       }
 
       VBK_LOG_DEBUG("Chain B wins");
+      ed_.overrideTip(candidate);
```
Comment (Warchant) [inline include/veriblock/pop/blockchain/pop/fork_resolution.hpp:666]: Also here

File: test/pop/blockchain/CMakeLists.txt [hunk 3]
```diff
@@ -68,4 +68,3 @@ addtest(temp_block_tree_test temp_block_tree_test.cpp)
 addtest(alt_tree_bootstrap_test alt_tree_bootstrap_test.cpp)
 addtest(altblock_finalization_test altblock_finalization_test.cpp)
 addtest(vbkblock_finalization_test vbkblock_finalization_test.cpp)
-addtest(altblock_restoration_test altblock_restoration_test.cpp)
```
Comment (Warchant) [inline test/pop/blockchain/CMakeLists.txt:71]: why you removed this test?

File: include/veriblock/pop/storage/util.hpp [hunk 4]
```diff
@@ -113,7 +113,7 @@ void saveTree(
   std::vector<const index_t*> dirty_indices;
 
   // map pair<hash, shared_ptr<index_t>> to vector<index_t*>
-  for (auto& index : tree.getAllBlocks()) {
+  for (auto& index : tree.getBlocks(false)) {
```
Comment (Warchant) [inline include/veriblock/pop/storage/util.hpp:116]: I think it would be better to have a enum class 
```
enum class GetBlocksPolicy {
  DEFAULT,
  SKIP_DELETED
}

```

and set getBlocks(GetBlocksPolicy policy = GetBlocksPolicy::DEFAULT)

File: include/veriblock/pop/blockchain/pop/fork_resolution.hpp (hunk 1 above)
Comment (blackyblack) [reply include/veriblock/pop/blockchain/pop/fork_resolution.hpp:555]: Because I want activateBestChain to be consistent. It should always switch to the winner.

File: test/pop/blockchain/CMakeLists.txt (hunk 3 above)
Comment (blackyblack) [reply test/pop/blockchain/CMakeLists.txt:71]: We never use block restoration code.

File: include/veriblock/pop/storage/util.hpp (hunk 4 above)
Comment (blackyblack) [reply include/veriblock/pop/storage/util.hpp:116]: A bit too verbose but I can do it this way.

Comment (blackyblack) [reply include/veriblock/pop/storage/util.hpp]: Done

File: include/veriblock/pop/blockchain/alt_block_tree.hpp [hunk 5]
```diff
@@ -115,19 +114,16 @@ extern template struct BaseBlockTree<AltBlock>;
...
+//! @note after AltBlockTree::activateBestChain AltBlockTree always corresponds to a state, as if winner chain have been applied.
 //!
 //!```cpp
 //!         if(result < 0) {
 //!             // candidate has better POP score.
 //!             // tree already switched to candidate chain.
 //!
-//!             UpdateTip(candidate->getHash());
-//!             // NOTE: update `tip`, otherwise old tip will be passed to first arg,
-//!             // and comparePopScore will die on assert
```
Comment (Warchant) [inline include/veriblock/pop/blockchain/alt_block_tree.hpp:130]: Why did you delete this?
Users must update tip of their block tree, otherwise ALT tree in alt-cpp and block tree in altchain won't be in sync

File: src/pop/blockchain/alt_block_tree.cpp [hunk 6]
```diff
@@ -546,8 +544,7 @@ void AltBlockTree::overrideTip(index_t& to) {
                  "the active chain tip(%s) must be fully valid",
                  to.toPrettyString());
 
-  onBeforeOverrideTip.emit(to);
```
Comment (Warchant) [inline src/pop/blockchain/alt_block_tree.cpp:549]: why did you remove this signal?

File: src/pop/blockchain/pop/vbk_block_tree.cpp [hunk 7]
```diff
@@ -35,28 +35,30 @@ void VbkBlockTree::determineBestChain(index_t& candidate,
...
-    this->overrideTip(candidate);
+    VBK_LOG_DEBUG("Candidate chain won");
   } else {
     // the current chain is better
     VBK_LOG_DEBUG("Active chain won");
   }
 }
 
+bool VbkBlockTree::setState(const hash_t& block, ValidationState& state) {
+  return base::setState(block, state);
```
Comment (Warchant) [inline src/pop/blockchain/pop/vbk_block_tree.cpp:54]: I think you meant to do this

```
auto* index = getBlockIndex(hash);
if (!index) return false;
return this->setState(*index, state);
```

**Review by Warchant — CHANGES_REQUESTED** [review]

File: src/pop/blockchain/pop/vbk_block_tree.cpp (hunk 7 above)
Comment (Warchant) [reply src/pop/blockchain/pop/vbk_block_tree.cpp:54]: Please add a testcase to reveal this problem

File: include/veriblock/pop/blockchain/alt_block_tree.hpp (hunk 5 above)
Comment (blackyblack) [reply include/veriblock/pop/blockchain/alt_block_tree.hpp:130]: I thought it is related to overrideTip call which is obsolete after this refactoring.

File: src/pop/blockchain/alt_block_tree.cpp (hunk 6 above)
Comment (blackyblack) [reply src/pop/blockchain/alt_block_tree.cpp:549]: It is not used anymore.

File: src/pop/blockchain/pop/vbk_block_tree.cpp (hunk 7 above)
Comment (blackyblack) [reply src/pop/blockchain/pop/vbk_block_tree.cpp:54]: It will call abstract base setState which looks for blockIndex by hash and redirects to specialized version of setState. I cannot make a test because there is no such issue.

Comment (Warchant) [reply src/pop/blockchain/pop/vbk_block_tree.cpp:54]: anyway, it is not clear from the code. please update

**Comment by Warchant** [comment]
Please rebase

**Comment by Warchant** [comment]
Copied everything except rename in #926
